### PR TITLE
fqdn: Avoid regenerations on each poller update

### PR DESCRIPTION
**Pull request by raybejjani**
This revives https://github.com/cilium/cilium/pull/6850 because I'm proposing https://github.com/cilium/cilium/pull/6997 as a 1.3 patch. If we merge one it would be good to merge the other.

When the DNS Poller would update DNS info we called StopPollDNSName
before calling StartPollDNSName. In cases where a new rule was added or
one removed, this was ok. In cases where we replaced an existing rule we
would remove a rule-uuid -> DNS name association, deleting the IPs for
that DNS name, and then add it again. This caused the update logic to
treat the IP update as new always, instead of only when the IPs had
changed (since it had no copy of the IP to compare to). This ultimately
resulted in a regeneration with every poller update.
The solution is to call StopPollDNSName on rules only after we have
called StartPollDNSName, and more aggresively reference count UUID-based
associations. This is especially important because we now randomize
UUIDs given to rules, and it was possible to not clean the UUID of an
old rule and regenerate it back into existence.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6998)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-docs-please

**Comment by ianvernon** [comment]
test-missed-k8s

**Review by tgraf — APPROVED** [review]
Looks good but I'm not super comfortable with this code

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
@raybejjani did you manually test that this worked? It looks fine at face value, but would be interested in seeing whether there were any before/after comparisons in v1.2 without / with this patch accordingly with respect to regenerations triggered by FQDN rule updates.

**Comment by ianvernon** [comment]
@eloycoto @nebril can you PTAL since you guys have worked with the ToFQDNs code ?

File: daemon/policy.go [hunk 1]
```diff
@@ -193,7 +193,7 @@ type AddOptions struct {
 	ReplaceWithLabels labels.LabelArray
 }
 
-func (d *Daemon) policyAdd(rules policyAPI.Rules, opts *AddOptions, prefixes []*net.IPNet) (uint64, error) {
+func (d *Daemon) policyAdd(rules policyAPI.Rules, opts *AddOptions, prefixes []*net.IPNet) (uint64, policyAPI.Rules, error) {
```
Comment (eloycoto) [inline daemon/policy.go:196]: Maybe a comment why the oldRules are needed in this function will help to get context in the future.

**Review by eloycoto — APPROVED** [review]

File: daemon/policy.go (hunk 1 above)
Comment (ianvernon) [reply daemon/policy.go:196]: I think the use of the `oldRules` and the explanation here suffices - https://github.com/cilium/cilium/pull/6998/files#diff-afacd8d4c41a906545f4ec8cde864b7fR295

**Pull request merged**